### PR TITLE
Ensure we wrap read with fastdom to prevent forced layout/style

### PR DESCRIPTION
**Pull request by nicl**
We are seeing forced layouts/styles because of email.js, specifically line 232. This PR aims to fix this by wrapping the read in fastdom.

screenshot 2019-01-31 at 15 25 31

(evidence of bad behaviour - note cumulative blocking time on a mobile device (here 4x CPU throttling in Chrome dev tools) can be 60ms+, which is not gigantic by any means, but still can block interactivity.

### CODE REVIEW

File: static/src/javascripts/projects/common/modules/email/email.js [hunk 1]
```diff
@@ -226,12 +226,13 @@ const setIframeHeight = (
...
-    fastdom.write(() => {
-        iFrameEl.height = '';
-        iFrameEl.height = `${
-            iFrameEl.contentWindow.document.body.clientHeight
-        }px`;
-        callback();
+    fastdom.read(() => {
+        const height = iFrameEl.contentWindow.document.body.clientHeight;
+
+        fastdom.write(() => {
```
Comment (SiAdcock) [inline static/src/javascripts/projects/common/modules/email/email.js]: Should we avoid calling `write` (mutate queue) from inside a `write` (measure queue)? In other places, we handle this like so:

```js
fastdom.read(() => iFrameEl.contentWindow.document.body.clientHeight)
  .then(height => {
    fastdom.write(() => {
      iFrameEl.height = `${height}px`;
      callback();
    });
  });
```

Can't find any specific docs on this!

Comment (nicl) [reply static/src/javascripts/projects/common/modules/email/email.js]: I think this is okay from the initial fastdom blog post and other examples:

screenshot 2019-01-31 at 15 49 35

Having said that, I don't know how deep the rabbit hole goes - i.e. can you just nest forever?!

Comment (SiAdcock) [reply static/src/javascripts/projects/common/modules/email/email.js]: Ah, actually we'd need to swap out `fastdom` for [`lib/fastdom-promise`](https://github.com/guardian/frontend/blob/master/static/src/javascripts/lib/fastdom-promise.js) in this module

Comment (nicl) [reply static/src/javascripts/projects/common/modules/email/email.js]: hey, so made this change. I've looked at https://github.com/wilsonpage/fastdom/blob/master/fastdom.js#L75 though and can't see any expensive operations (it adds stuff to an array, and does a few conditionals and schedules a raf callback) so suspect this is unlikely to be significant in practice.

Comment (SiAdcock) [reply static/src/javascripts/projects/common/modules/email/email.js]: Yeah, I think you're right, it shouldn't make any difference

File: static/src/javascripts/lib/fastdom-promise.js [hunk 2]
```diff
@@ -22,4 +22,5 @@ const promisify = fdaction => (fn: Function, ctx: ?Object): Promise<any> =>
 export default {
     read: promisify(fastdom.read.bind(fastdom)),
     write: promisify(fastdom.write.bind(fastdom)),
+    defer: promisify(fastdom.defer.bind(fastdom)),
```
Comment (nicl) [inline static/src/javascripts/lib/fastdom-promise.js]: (This is used in email.js.)

**Review by SiAdcock — APPROVED** [review]
I think this is fine, but it was probably equally fine before my unnecessary intervention! Nice work 👍

**Pull request merged**